streams/shared/src/main/scala/zio/stream/ZSummary.scala

```diff
@@ -0,0 +1,813 @@
+                }
+            }
+          }
+      }
+
+    }
+
+  /**
+   * Applies a function to this summary's result.
+   */
+  def map[O2](f: O => O2): ZSummary[R, E, I, O2] =
+    mapBuilder {
+      case builder: Builder.Mutable[I, O] =>
+        def wrap(orig: Builder.Mutable[I, O]): Builder.Mutable[I, O2] =
+          new Builder.Mutable[I, O2] {
+            override def accepts(i: I): Boolean         = orig.accepts(i)
+            override def copy(): Builder.Mutable[I, O2] = wrap(orig.copy())
+            override def done(): Boolean                = orig.done()
+            override def reset(): Unit                  = orig.reset()
+            override def step(i: I): Unit               = orig.step(i)
+            override def extract(): O2                  = f(orig.extract())
+          }
+
+        wrap(builder)
+
+      case builder: Builder.Effectual[R, E, I, O] =>
+        new Builder.Effectual[R, E, I, O2] {
+          def accepts(i: I): ZIO[R, E, Boolean] = builder.accepts(i)
+          def reset: ZIO[R, E, Unit]            = builder.reset
+          def done: ZIO[R, E, Boolean]          = builder.done
+          def step(i: I): ZIO[R, E, Unit]       = builder.step(i)
+          def extract: ZIO[R, E, O2]            = builder.extract.map(f)
+        }
+    }
+
+  /**
+   * Applies an effectual function to this summary's result.
+   */
+  def mapM[R1 <: R, E1 >: E, O2](f: O => ZIO[R1, E1, O2]): ZSummary[R1, E1, I, O2] =
+    mapBuilder {
+      case builder: Builder.Mutable[I, O] =>
+        new Builder.Effectual[R1, E1, I, O2] {
+          def reset: ZIO[R1, E1, Unit]            = UIO(builder.reset())
+          def accepts(i: I): ZIO[R1, E1, Boolean] = UIO(builder.accepts(i))
+          def done: ZIO[R1, E1, Boolean]          = UIO(builder.done())
+          def step(i: I): ZIO[R1, E1, Unit]       = UIO(builder.step(i))
+          def extract: ZIO[R1, E1, O2]            = UIO(builder.extract()).flatMap(f)
+        }
+
+      case builder: Builder.Effectual[R, E, I, O] =>
+        new Builder.Effectual[R1, E1, I, O2] {
+          def reset: ZIO[R1, E1, Unit]            = builder.reset
+          def accepts(i: I): ZIO[R1, E1, Boolean] = builder.accepts(i)
+          def done: ZIO[R1, E1, Boolean]          = builder.done
+          def step(i: I): ZIO[R1, E1, Unit]       = builder.step(i)
+          def extract: ZIO[R1, E1, O2]            = builder.extract.flatMap(f)
+        }
+    }
+
+  def mapBuilder[R1, E1, I1, O1](f: Builder[R, E, I, O] => Builder[R1, E1, I1, O1]): ZSummary[R1, E1, I1, O1] =
+    ZSummary(self.builder.map(f))
+
+  /**
+   * Creates a transducer from this summary
+   */
+  def toTransducer: ZTransducer[R, E, I, O] =
+    ZTransducer.fromSummary(self)
+
+  /**
+   * Creates a sink from this summary.
+   */
+  def toSink[I1 <: I]: ZSink[R, E, I1, I1, O] =
+    ZSink.fromSummary(self)
+
+  private def zipMutableEffectual[R0, E0, I0, A, B, Z](
```

Comment:
Do we have a convention of putting private methods after public ones?

Reply:
:shrug: happy to adjust